---
title: "Identifying Copy Number Polymorphisms"
author: "Jacob Carey, Steven Cristiano, and Robert Scharpf"
date: \today
output: BiocStyle::pdf_document
vignette: >
  %\VignetteIndexEntry{Identifying Copy Number Polymorphisms}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc} 
---

# Introduction

Identify consensus start and stop coordinates of a copy number
polymorphism

The collection of copy number variants (CNVs) identified in a study
can be encapulated in a GRangesList, where each element is a GRanges
of the CNVs identified for an individual.  (For a study with 1000
subjects, the GRangesList object would have length 1000 if each
individual had 1 or more CNVs.)  For regions in which CNVs occur in
more than 2 percent of study participants, the start and end
boundaries of the CNVs may differ because of biological differences in
the CNV size as well as due to technical noise of the assay and the
uncertainty of the breakpoints identified by a segmentation of the
genomic data.  Among subjects with a CNV called at a given locus, the
`consensusCNP` function identifies the largest region that is copy
number variant in half of these subjects.

# Finding CNPs
- hidden Markov models
- prior knowledge (HapMap / 1000 Genomes Project)

# Outline from Rob
Preliminary

RS: help!!
```{r prelim}
knitr::opts_chunk$set(eval=FALSE)
library(CNPBayes)
library(VanillaICE)
# library(pd.genomewidesnp.6)
# genome <- pd.genomewidesnp.6
 
# change this to include subset of SNP
# have set of SNPs with log r ratios and then summarize in some
# way. use some of the data from the vanilla ices vignette and then
# use that here. has SNP data for chromosome and we just have to
                                        # simulate more of it.

library(BSgenome.Hsapiens.UCSC.hg19)
genome <- BSgenome.Hsapiens.UCSC.hg19

set.seed(100)
starts <- rpois(1000, 100) + 10e6L
ends <- rpois(1000, 100) + 10.1e6L
cnv1 <- GRanges("chr1", IRanges(starts, ends))
cnv1$id <- paste0("sample", seq_along(cnv1))
seqinfo(cnv1) <- seqinfo(genome)

starts <- rpois(500, 1000) + 101e6L
ends <- rpois(500, 1000) + 101.4e6L
cnv2 <- GRanges("chr5", IRanges(starts, ends))
cnv2$id <- paste0("sample", seq_along(cnv2))
seqlevels(cnv2, force=TRUE) <- seqlevels(genome)
seqinfo(cnv2) <- seqinfo(genome)

all_cnvs <- c(cnv1, cnv2)
```

Step 1

```{r step-1}
cnps <- consensusCNP(all_cnvs)
```

Step 2
```{r step-2}
create a GRanges object for the markers
affy 6.0
- find annotation for SNPs and nonpolymorphic markers

features.granges - create a GRanges object
```

Step 3 
```{r step-3}
findOverlaps(cnps, features)
            query  subject
```

Step 4
```{r step-4}
SummarizedExperiment w/ findOverlaps
2 assays: log R ratios, BAFs
```

Step 5 
```{r step-5}
create a 1d summary of the log r ratios 
mention PCA, but use mean
```

Step 6
```{r step-6}
SummarizedExperiment for the 1-d summary.
rowRanges: CNPs in step 1
```

Step 7
```{r step-7}
Fit the `MixtureModel` to each row of the object in step 6.
```
